### ISSUE TITLE
Do not ignore unknown values in {flags}/{options} arguments

### ISSUE THREAD

**Issue by ZyX-I** [vim/vim#2858]
Currently if `{flags}` string or `{options}` dictionary contain unknown value it is simply ignored. This behaviour is wrong on at least two accounts:

- It does not allow simple and straightforward way to check whether some functionality is supported (i.e. try to use some flag/option and catch the failure).
- It may result in plugin corrupting data or malfunctioning without any visible errors.

One of the examples is relatively recent addition of `a` flag to `writefile()`: if plugin just uses it without caring whether it is supported it at best could make some problem not easily debuggable (by failing to save non-last log entries) and at worst may provoke data loss or corruption (if used to create a data file).

I initially opened issue in Neovim, but after collecting this [nice estimation of ability to ignore *existing* flags in various functions](https://github.com/neovim/neovim/issues/8284#issuecomment-385808785) I think it is good idea to open it here too.

**Comment by k-takata** [comment]
When I read this first time, the meaning of {flags}/{options} was unclear to me.
After reading the link, I understand that they are options for each Vim script function.

The problem is that this is not backward compatible and might break existing plugins, right?

**Comment by jamessan** [comment]
> The problem is that this is not backward compatible and might break existing plugins, right?

No.  Plugins are already broken, but they don't know it.  Fixing this would make it so the brokenness is easily discoverable, due to Vim raising errors, rather than causing subtle issues that have to be investigated.

**Comment by ZyX-I** [comment]
@jamessan This is the case for flags and options of those functions who do not take callbacks in `{options}` and then use the very same `{options}` dictionary as `self`. Those would be not broken, but if `{options}` is used as `self` there are good chances something break. Apparently, based on Vim doc stating that plugin author needs to bind `self` himself (`:h channel-open-options`, callback option) and behaviour of `job_start()` which is not stated in documentation, but is basically the same it is mostly the case for Neovim. I for some reason thought Bram took the same approach, but it appears not the case so something will break like this only if plugin contains weird code making sure that `{options}` is `self` and storing some unrelated data there.

If I recall course of events correctly this is the result of partials appearing in Vim *earlier* then jobs, while it is the opposite in Neovim.

**Comment by brammool** [comment]
Nikolai Pavlov wrote:

> Currently if `{flags}` string or `{options}` dictionary contain unknown value it is simply ignored. This behaviour is wrong on at least two accounts:
> 
> - It does not allow simple and straightforward way to check whether some functionality is supported (i.e. try to use some flag/option and catch the failure).
> - It may result in plugin corrupting data or malfunctioning without any visible errors.
> 
> One of the examples is relatively recent addition of `a` flag to `writefile()`: if plugin just uses it without caring whether it is supported it at best could make some problem not easily debuggable (by failing to save non-last log entries) and at worst may provoke data loss or corruption (if used to create a data file).
> 
> I initially opened issue in Neovim, but after collecting this [nice estimation of ability to ignore *existing* flags in various functions](https://github.com/neovim/neovim/issues/8284#issuecomment-385808785) I think it is good idea to open it here too.

In some cases adding unsupported flags or options do give an error.
But in many cases they don't.  The main reason appears to be that we
have been lazy adding checks...

In general it's good to give an error, both for detecting using an older
version where the feature is not supported, and also to catch typos when
writing Vim script.

-- 
"A mouse can be just as dangerous as a bullet or a bomb."
             (US Representative Lamar Smith, R-Texas)

 /// Bram Moolenaar -- Bram@Moolenaar.net -- http://www.Moolenaar.net   \\\
///        sponsor Vim, vote for features -- http://www.Vim.org/sponsor/ \\\
\\\  an exciting new programming language -- http://www.Zimbu.org        ///
 \\\            help me help AIDS victims -- http://ICCF-Holland.org    ///
